src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -134,120 +147,171 @@ private async Task StartShardSubscriptionWorkers()
 
         private async Task MaintainConnectionWithClientWorker()
         {
-            var exceptions = new Dictionary<string, Exception>();
-            var shardsToReconnect = new List<string>();
-            var heartbeatSp = Stopwatch.StartNew();
-            var reconnectSp = Stopwatch.StartNew();
-            var didBatch = false;
+            var hasBatch = _mre.WaitAsync(CancellationTokenSource.Token);
```

Comment:
Why do it here and not in the loop?

Reply:
I don't want to create a new task each time we start the loop, we only can recreate it when the task complete (which mean that some shard worker set the `_mre`)